### PR TITLE
feat: WEB-6859 Declarative depends on

### PR DESCRIPTION
**Pull request by nick4u**
related -> WEB-6859

### CODE REVIEW

**Review by wiktorwojcikowski-flyr — APPROVED** [review]

File: src/alembic_utils/replaceable_entity.py [hunk 1]
```diff
@@ -173,15 +174,45 @@ def get_required_migration_op(
...
+    expanded_deps = set()
+    for e in entity.depends_on:
+        expanded_deps.add(e)
+        expanded_deps.update(depends_on_expanded(e))
+
+    return expanded_deps
+
+
+def dependants(entities: list[ReplaceableEntity], known_obj_with_deps: list[ReplaceableEntity]) -> set[ReplaceableEntity]:
+    """return set of objects that directly on indirectly depend on any entity in entities"""
```
Comment (j-janowski) [inline src/alembic_utils/replaceable_entity.py]:
```suggestion
    """return set of objects that directly or indirectly depend on any entity in entities"""
```

File: src/alembic_utils/replaceable_entity.py [hunk 2]
```diff
@@ -310,6 +353,26 @@ def compare_registered_entities(
...
+        for drop in sorted(depends_on_changed, key=ordered_entities.index):
+            if drop.identity in [le.identity for le in local_entities]:
+                upgrade_ops.ops.insert(0, DropOp(drop))
+            else:
+                logger.debug(f"drop ignore for new object: {drop.identity}")
+
+        # apply changes found
+        upgrade_ops.ops.extend(e.op for e in changed_entities)
+
+        # recreate (crate) objects that depend on objects dropped before
```
Comment (j-janowski) [inline src/alembic_utils/replaceable_entity.py]:
```suggestion
        # recreate (create) objects that depend on objects dropped before
```

File: src/test/test_depends_on.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+def test_create_revision_with_explicit_depends_on(engine) -> None:
+    register_entities([B_A, E_AD, D_B, C_A, A])
+
+    run_alembic_command(
+        engine=engine,
+        command="revision",
+        command_kwargs={"autogenerate": True, "rev_id": "1", "message": "create"},
+    )
+    print(run_alembic_command(engine=engine, command="upgrade", command_kwargs={"revision": "1", "sql": True}))
```
Comment (j-janowski) [inline src/test/test_depends_on.py]: I can see in this test file multiple debug prints

File: src/alembic_utils/replaceable_entity.py [hunk 4]
```diff
@@ -173,15 +174,45 @@ def get_required_migration_op(
         return CreateOp(self)
 
 
+class EntityWithOp(NamedTuple):
+    entity: ReplaceableEntity
+    op: ReversibleOp
+
+
+def depends_on_expanded(entity: ReplaceableEntity) -> set[ReplaceableEntity]:
```
Comment (j-janowski) [inline src/alembic_utils/replaceable_entity.py:182]: I think we need some unit tests to cover this function.

File: src/alembic_utils/replaceable_entity.py [hunk 5]
```diff
@@ -173,15 +174,45 @@ def get_required_migration_op(
...
+    """
+    expanded_deps = set()
+    for e in entity.depends_on:
+        expanded_deps.add(e)
+        expanded_deps.update(depends_on_expanded(e))
+
+    return expanded_deps
+
+
+def dependants(entities: list[ReplaceableEntity], known_obj_with_deps: list[ReplaceableEntity]) -> set[ReplaceableEntity]:
```
Comment (j-janowski) [inline src/alembic_utils/replaceable_entity.py:194]: Same as in case of `depends_on_expanded`, dedicated ut would be great

**Review by j-janowski — COMMENTED** [review]
Great work! I added some small comments and suggestions which can smooth this PR if we plan to contribute it to open source.

File: src/alembic_utils/replaceable_entity.py (hunk 4 above)
Comment (nick4u) [reply src/alembic_utils/replaceable_entity.py:182]: added

File: src/alembic_utils/replaceable_entity.py (hunk 5 above)
Comment (nick4u) [reply src/alembic_utils/replaceable_entity.py:194]: added

**Review by j-janowski — APPROVED** [review]

File: src/test/test_depends_on.py (hunk 3 above)
Comment (nick4u) [reply src/test/test_depends_on.py]: As agreed left in commented state for now

**Review by wrzasa — APPROVED** [review]

**Pull request merged**